crates/wasmparser/src/readers/core/types.rs

```diff
@@ -762,6 +881,75 @@ impl FuncType {
     }
 }
 
+impl Matches for FuncType {
+    fn matches<'a, F>(&self, other: &Self, type_at: &F) -> Result<bool>
+    where
+        F: Fn(u32) -> Result<&'a SubType>,
+    {
+        let r = self.params().len() == other.params().len()
+            && self.results().len() == other.results().len()
+            // Note: matching functions are contravariant in their parameter types.
+            && self
+                .params()
+                .iter()
+                .zip(other.params())
+                .try_fold(true, |r, (a, b)| Ok(r && b.matches(a, type_at)?))?
+            && self
+                .results()
+                .iter()
+                .zip(other.results())
+                .try_fold(true, |r, (a, b)| Ok(r && a.matches(b, type_at)?))?;
+        Ok(r)
+    }
+}
+
+impl Matches for ArrayType {
+    fn matches<'a, F>(&self, other: &Self, type_at: &F) -> Result<bool>
+    where
+        F: Fn(u32) -> Result<&'a SubType>,
+    {
+        self.0.matches(&other.0, type_at)
+    }
+}
+
+impl Matches for FieldType {
+    fn matches<'a, F>(&self, other: &Self, type_at: &F) -> Result<bool>
+    where
+        F: Fn(u32) -> Result<&'a SubType>,
+    {
+        Ok((other.mutable || !self.mutable)
+            && self.element_type.matches(&other.element_type, type_at)?)
+    }
+}
+
+impl Matches for StorageType {
+    fn matches<'a, F>(&self, other: &Self, type_at: &F) -> Result<bool>
+    where
+        F: Fn(u32) -> Result<&'a SubType>,
+    {
+        match (self, other) {
+            (Self::Val(a), Self::Val(b)) => a.matches(b, type_at),
+            (a, b) => Ok(*a == *b),
```

Comment:
Another catch-all I think would be good to remove

Reply:
Removed.